python/hail/experimental/ldscore.py

```diff
@@ -71,107 +69,88 @@ def ld_score(entry_expr, annotation_exprs,
-    n_positions = len(positions)
-
-    starts = np.zeros(n_positions, dtype='int')
-    stops = np.zeros(n_positions, dtype='int')
-
-    contig = '0'
-    for i, (c, p) in enumerate(positions):
-        if c != contig:
-            j = i
-            k = i
-            contig = c
-
-        min_val = p - window_size
-        max_val = p + window_size
-
-        while j < n_positions and positions[j][1] < min_val:
-            j += 1
-
-        starts[i] = j
-
-        if k == n_positions:
-            stops[i] = k
-            continue
-
-        while positions[k][0] == contig and positions[k][1] <= max_val:
-            k += 1
-            if k == n_positions:
-                break
-
-        stops[i] = k
-
-    R2_adj_sparse = R2_adj.sparsify_row_intervals([int(x) for x in starts],
-                                                  [int(x) for x in stops])
-    L2 = R2_adj_sparse @ A
-
-    tmp_bm_path = new_temp_file()
-    tmp_tsv_path = new_temp_file()
-
-    L2.write(tmp_bm_path, force_row_major=True)
-    BlockMatrix.export(tmp_bm_path, tmp_tsv_path)
-
-    ht_scores = hl.import_table(tmp_tsv_path, no_header=True, impute=True)
+    N = mt.count_cols()
+
+    R2 = hl.row_correlation(entry_expr, block_size) ** 2
+    R2_adj = (N - 1.0)/(N - 2.0) * R2 - 1.0/(N - 2.0)
+
+    starts, stops = hl.linalg.utils.locus_windows(locus_expr,
+                                                  radius,
+                                                  coord_expr)
+    R2_adj_sparse = R2_adj.sparsify_row_intervals(starts, stops)
+
+    tmp_R2_adj_sparse_file = new_temp_file()
+    R2_adj_sparse.write(tmp_R2_adj_sparse_file, force_row_major=True)
+    R2_adj_sparse = BlockMatrix.read(tmp_R2_adj_sparse_file)
+
+    if not annotation_exprs:
+        cols = ['univariate']
+        col_idxs = {0: 'univariate'}
+        L2 = R2_adj_sparse.sum(axis=1)
+    else:
+        ht = mt.select_rows(*wrap_to_list(annotation_exprs)).rows()
+        ht = ht.annotate(univariate=hl.literal(1.0))
+        names = [x for x in ht.row if x not in ht.key]
+
+        ht_union = hl.Table.union(
+            *[(ht.annotate(name=hl.str(x),
+                           value=hl.float(ht[x]))
+                 .select('name', 'value')) for x in names])
+        mt_annotations = ht_union.to_matrix_table(
+            row_key=[x for x in ht_union.key],
+            col_key=['name'])
+
+        cols = mt_annotations['name'].collect()
+        col_idxs = {i: cols[i] for i in range(len(cols))}
+
+        A = BlockMatrix.from_entry_expr(mt_annotations.value)
+        tmp_A_file = new_temp_file()
+        BlockMatrix.write_from_entry_expr(mt_annotations.value, tmp_A_file)
+
+        A = BlockMatrix.read(tmp_A_file)
+        L2 = R2_adj_sparse @ A
+
+    tmp_L2_bm_file = new_temp_file()
+    tmp_L2_tsv_file = new_temp_file()
+    L2.write(tmp_L2_bm_file, force_row_major=True)
+    BlockMatrix.export(tmp_L2_bm_file, tmp_L2_tsv_file)
```

Comment:
how big is this in practice?

Reply:
A standard set of SNPs would be ~1.2M HapMap3 SNPs. Kate would like to use on the order of ~20,000 annotations.